core/src/main/java/software/amazon/awssdk/retry/RetryHandler.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2010-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.retry;
+
+import static java.util.Collections.singletonList;
+
+import java.util.Optional;
+import software.amazon.awssdk.RequestExecutionContext;
+import software.amazon.awssdk.SdkBaseException;
+import software.amazon.awssdk.http.HttpResponse;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+import software.amazon.awssdk.retry.v2.RetryPolicy;
+import software.amazon.awssdk.retry.v2.RetryPolicyContext;
+import software.amazon.awssdk.util.CapacityManager;
+
+public class RetryHandler {
```

Comment:
Is this class meant to be shared between threads?

Reply:
Not really. This class is going to be instantiated on each request / retry and its state is created newly based on the data at hand.